Datasources - Part 2: main extension + existStrict + builder

rolnico opened the pull request:
**Please check if the PR fulfills these requirements**
<!-- please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes -->
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**Does this PR already have an issue describing the problem?**
<!-- If so, link to this issue using `'Fixes #XXX'` and skip the rest -->



**What kind of change does this PR introduce?**
<!-- Bug fix, feature, docs update, ... -->



**What is the current behavior?**
<!-- You can also link to an open issue here -->



**What is the new behavior (if this is a feature change)?**



**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**
<!-- If no breaking changes or API deprecations were introduced, delete this section -->
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**
<!-- If this PR introduces a breaking change, describe the migration steps -->
<!-- The content of this section will be copied in the migration guide -->



**Other information**:
<!-- if any of the questions/checkboxes don't apply, please delete them entirely -->

jonenst commented on commons/src/main/java/com/powsybl/commons/datasource/DirectoryDataSource.java:
```diff
@@ -59,7 +62,8 @@ protected OutputStream getCompressedOutputStream(OutputStream os) throws IOExcep
         Objects.requireNonNull(fileName);
-        return directory.resolve(fileName + getCompressionExt());
+        FileInformation fileInformation = new FileInformation(fileName, false);
+        return directory.resolve(fileInformation.getCompressionFormat() == null ? fileName + getCompressionExtension() : fileName);
```
mention this fix/change or put in a separate PR ?

This changes compressed directory data source to allow newinputstream newoutputstream and exists to work with "foo.xiidm" when the file is "foo.xiidm.gz"

so it introduces the concept of "filenameOrUncompressedFilename" ?

rolnico replied:
added info about this in the PR comment + in the breaking change section

jonenst commented on iidm/iidm-serde/src/test/java/com/powsybl/iidm/serde/ConversionTest.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @Test
+    void testConversionZip() throws IOException {
```
very nice that you added a test !

jonenst commented:
since everything can be implemented only with getBaseExtension ( see #3085 ) , do we still want to add an API existsStrict ? Maybe yes but i'm not sure

rolnico commented:
> since everything can be implemented only with getBaseExtension ( see #3085 ) , do we still want to add an API existsStrict ? Maybe yes but i'm not sure

I find it better to do the check on the datasource side and not on the importer side. However, after discussion with @flo-dup, I changed `existsStrict(String suffix, String ext)` to `isMainExtension(String ext)` and fixed the uses. This method is more understandable and still keeps the logic on the datasource side.

flo-dup commented on commons/src/main/java/com/powsybl/commons/datasource/FileInformation.java:
```diff
@@ -0,0 +1,115 @@
+        int currentDotIndex = fileName.lastIndexOf('.');
+
+        // Compression extension
+        compressionFormat = switch (fileName.substring(currentDotIndex + 1)) {
```
My bad! `s.substring(s.length())` does not throw an exception but returns an empty string!!!

flo-dup commented on commons/src/main/java/com/powsybl/commons/datasource/FileInformation.java:
```diff
@@ -0,0 +1,115 @@
+    private String baseName;
+    private CompressionFormat compressionFormat;
+    private ArchiveFormat archiveFormat;
+    private String mainExtension;
```
ok you named it mainExtension in opposition with the compression extension, got it

flo-dup commented on commons/src/main/java/com/powsybl/commons/datasource/ReadOnlyDataSource.java:
```diff
@@ -29,11 +33,18 @@ public interface ReadOnlyDataSource {
     /**
      * Check if a file exists in the datasource.
-     * @param fileName Name of the file (excluding the compression extension)
+     * @param fileName Name of the file
```
In fact this comment should rather be in `AbstractFileSystemDataSource` as compression doesn't exist at this stage.

jonenst commented on commons/src/main/java/com/powsybl/commons/datasource/ReadOnlyDataSource.java:
```diff
@@ -34,6 +38,13 @@ public interface ReadOnlyDataSource {
+     * @param ext Extension to compare to the datasource main extension
+     * @return true if the datasource main extension is null, empty or equal to {@code ext}, else false
+     */
+    boolean isMainExtension(String ext);
```
For cgmes, because the filename can be anything, you can't just verify existsAsMainFile for one known filename (basename.ext)

flo-dup replied:
ok to keep it as it is now. Let's find a suitable refactor later as we didn't find a common agreement these days, and this shouldn't block this PR.

rolnico commented on cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesOnDataSource.java:
```diff
@@ -24,6 +25,8 @@
  * @author Luma Zamarreño {@literal <zamarrenolm at aia.es>}
  */
 public class CgmesOnDataSource {
+    private static final String[] EXTENSIONS = {"xml"};
```
to be consistent with the importers

rolnico replied:
but that can be changed, if needed

jonenst commented on cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesOnDataSource.java:
```diff
@@ -32,29 +34,55 @@ public ReadOnlyDataSource dataSource() {
+        if (dataSource.getDataExtension() == null || dataSource.getDataExtension().isEmpty()) {
+            return false;
+        } else if (EXTENSIONS.equals(dataSource.getDataExtension())) {
+            try (InputStream is = dataSource.newInputStream(null, EXTENSIONS)) {
```
Does this do an NPE if the user specified a data extension but the file doesn't actually exist ?

rolnico replied:
It would throw an exception but not a NPE. It will be fixed in https://github.com/powsybl/powsybl-core/pull/3114
